src/client-side-encryption/state_machine.ts

```diff
@@ -156,14 +157,15 @@ export class StateMachine {
    */
   async execute<T extends Document>(
     executor: StateMachineExecutable,
-    context: MongoCryptContext
+    context: MongoCryptContext,
+    responseType?: MongoDBResponseConstructor
   ): Promise<T> {
     const keyVaultNamespace = executor._keyVaultNamespace;
     const keyVaultClient = executor._keyVaultClient;
     const metaDataClient = executor._metaDataClient;
     const mongocryptdClient = executor._mongocryptdClient;
     const mongocryptdManager = executor._mongocryptdManager;
-    let result: T | null = null;
+    let result: any | null = null;
```

Comment:
I mean, TypeScript would be kind of right to complain here? The return type doesn't necessarily `extends Document` anymore

I think for "proper" typing you'd need to have overloads that account for the presence or absence of `responseType`:

```ts
    async execute<R extends typeof MongoDBResponseConstructor & {new(): R['prototype']}>(..., responseType: R): Promise<R['prototype']>;
    async execute<T extends Document>(...): Promise<T>;
    async execute<T extends Document, R extends typeof MongoDBResponseConstructor & {new(): R['prototype']}>(..., responseType?: R): Promise<T | R['prototype']> {
```

([TS playground](https://www.typescriptlang.org/play?ssl=14&ssc=158&pln=11&pc=12#code/JYOwLgpgTgZghgYwgAgCIHsEFcC2FzIDeAUMmcgNoDWEAngFzIDOYUoA5gLqMtsjsBuYgF9ixUJFiIUAFXQAHADIQAbhAA2GbHgIQAHpBAATJmky58YIqXIB9YEZ6sOQ0cQTq4TUwFl0-dFQAIQAlCCZ5fyYIAGEo1iwEMHQoIjcPL1MY9VMScmQvWhAEZH0IbEgAHhDSg3wTZDBaeQh0GGQ-AOCwiKjY+KhE5NSAMiIQCAB3AAoASkYQigByeSh0ZKaWpc5hAD5pqHDIkGiZZogF+eQABTWcYGjq5dX19fPt3aF8wuLa8qwqjJaoYGloLOB9ldbuh7o8ZJ8bGQfiUyhUIJUgfoQaYwTowAAaZA1LH1UybVrtTrsQKhI59OInBJJFLIMaECYzK6LFZrDbvHb7Q69E4QM4tAD8l0Y0Nh6KBAB8ic9eW8tpxdtZ8uRgO0DnSRWKILNkIcwFgoCBkByTfrTuc5l8tTazRaiDB0OhGEsAEZwKBLYSOsiidLxZCgFhwYooAC8VqmyGyTAd7jD7vQyDjEbAUaQADpUQCINMqTSesdogzeEMUrMhAgw77UlnGbmIAW9P9ICmG4zkL6AF6Z8Ot6MdrtyhTKNSacx4yFCIA), wanted to check if this actually works 🙂 )